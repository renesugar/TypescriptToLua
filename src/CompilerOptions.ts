import * as ts from "typescript";

export interface CompilerOptions extends ts.CompilerOptions {
    noHeader?: boolean;
    luaTarget?: LuaTarget;
    luaLibImport?: LuaLibImportKind;
    noHoisting?: boolean;
}

export enum LuaLibImportKind {
    None = "none",
    Always = "always",
    Inline = "inline",
    Require = "require",
}

export enum LuaTarget {
    Lua51 = "5.1",
    Lua52 = "5.2",
    Lua53 = "5.3",
    LuaJIT = "jit",
}
