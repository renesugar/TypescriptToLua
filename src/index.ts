export {parseCommandLine} from "./CommandLineParser";
export {compile, compileFilesWithOptions, transpileString, watchWithOptions} from "./Compiler";
export {CompilerOptions, LuaLibImportKind, LuaTarget,} from "./CompilerOptions";
export {LuaLibFeature,} from "./LuaLib";
export {LuaTranspiler,} from "./LuaTranspiler";
