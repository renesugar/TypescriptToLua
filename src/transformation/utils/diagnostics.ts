import * as ts from "typescript";
import { LuaTarget } from "../../CompilerOptions";
import { createSerialDiagnosticFactory } from "../../utils";
import { AnnotationKind } from "./annotations";

const createDiagnosticFactory = <TArgs extends any[]>(message: string | ((...args: TArgs) => string)) =>
    createSerialDiagnosticFactory((node: ts.Node, ...args: TArgs) => ({
        file: node.getSourceFile(),
        start: node.getStart(),
        length: node.getWidth(),
        messageText: typeof message === "string" ? message : message(...args),
    }));

export const unsupportedNodeKind = createDiagnosticFactory(
    (kind: ts.SyntaxKind) => `Unsupported node kind ${ts.SyntaxKind[kind]}`
);

export const forbiddenForIn = createDiagnosticFactory("Iterating over arrays with 'for ... in' is not allowed.");

export const unsupportedNoSelfFunctionConversion = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return (
        `Unable to convert function with a 'this' parameter to function${nameReference} with no 'this'. ` +
        "To fix, wrap in an arrow function, or declare with 'this: void'."
    );
});

export const unsupportedSelfFunctionConversion = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return (
        `Unable to convert function with no 'this' parameter to function${nameReference} with 'this'. ` +
        "To fix, wrap in an arrow function, or declare with 'this: any'."
    );
});

export const unsupportedOverloadAssignment = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` to '${name}'` : "";
    return (
        `Unsupported assignment of function with different overloaded types for 'this'${nameReference}. ` +
        "Overloads should all have the same type for 'this'."
    );
});

export const decoratorInvalidContext = createDiagnosticFactory("Decorator function cannot have 'this: void'.");

export const annotationInvalidArgumentCount = createDiagnosticFactory(
    (kind: AnnotationKind, got: number, expected: number) => `'@${kind}' expects ${expected} arguments, but got ${got}.`
);

export const extensionCannotConstruct = createDiagnosticFactory(
    "Cannot construct classes with '@extension' or '@metaExtension' annotation."
);

export const extensionCannotExtend = createDiagnosticFactory(
    "Cannot extend classes with '@extension' or '@metaExtension' annotation."
);

export const extensionCannotExport = createDiagnosticFactory(
    "Cannot export classes with '@extension' or '@metaExtension' annotation."
);

export const extensionInvalidInstanceOf = createDiagnosticFactory(
    "Cannot use instanceof on classes with '@extension' or '@metaExtension' annotation."
);

export const extensionAndMetaExtensionConflict = createDiagnosticFactory(
    "Cannot use both '@extension' and '@metaExtension' annotations on the same class."
);

export const metaExtensionMissingExtends = createDiagnosticFactory(
    "'@metaExtension' annotation requires the extension of the metatable class."
);

export const invalidForRangeCall = createDiagnosticFactory((message: string) => `Invalid @forRange call: ${message}.`);

export const luaTableMustBeAmbient = createDiagnosticFactory(
    "Classes with the '@luaTable' annotation must be ambient."
);

export const luaTableCannotBeExtended = createDiagnosticFactory(
    "Cannot extend classes with the '@luaTable' annotation."
);

export const luaTableInvalidInstanceOf = createDiagnosticFactory(
    "The instanceof operator cannot be used with a '@luaTable' class."
);

export const luaTableCannotBeAccessedDynamically = createDiagnosticFactory("@luaTable cannot be accessed dynamically.");

export const luaTableForbiddenUsage = createDiagnosticFactory(
    (description: string) => `Invalid @luaTable usage: ${description}.`
);

export const luaIteratorForbiddenUsage = createDiagnosticFactory(
    "Unsupported use of lua iterator with '@tupleReturn' annotation in for...of statement. " +
        "You must use a destructuring statement to catch results from a lua iterator with " +
        "the '@tupleReturn' annotation."
);

export const unsupportedAccessorInObjectLiteral = createDiagnosticFactory(
    "Accessors in object literal are not supported."
);

export const unsupportedRightShiftOperator = createDiagnosticFactory(
    "Right shift operator is not supported for target Lua 5.3. Use `>>>` instead."
);

const getLuaTargetName = (version: LuaTarget) => (version === LuaTarget.LuaJIT ? "LuaJIT" : `Lua ${version}`);
export const unsupportedForTarget = createDiagnosticFactory(
    (functionality: string, version: Exclude<LuaTarget, LuaTarget.Universal>) =>
        `${functionality} is/are not supported for target ${getLuaTargetName(version)}.`
);

export const unsupportedProperty = createDiagnosticFactory(
    (parentName: string, property: string) => `${parentName}.${property} is unsupported.`
);

export const invalidAmbientIdentifierName = createDiagnosticFactory(
    (text: string) => `Invalid ambient identifier name '${text}'. Ambient identifiers must be valid lua identifiers.`
);

export const unresolvableRequirePath = createDiagnosticFactory(
    (path: string) => `Cannot create require path. Module '${path}' does not exist within --rootDir.`
);

export const unsupportedVarDeclaration = createDiagnosticFactory(
    "`var` declarations are not supported. Use `let` or `const` instead."
);
