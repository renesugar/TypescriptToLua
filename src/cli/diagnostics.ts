import * as ts from "typescript";
import { createSerialDiagnosticFactory, createDiagnosticFactoryWithCode } from "../utils";

export const tstlOptionsAreMovingToTheTstlObject = createSerialDiagnosticFactory((tstl: Record<string, any>) => ({
    category: ts.DiagnosticCategory.Warning,
    messageText:
        'TSTL options are moving to the "tstl" object. Adjust your tsconfig to look like\n' +
        `"tstl": ${JSON.stringify(tstl, undefined, 4)}`,
}));

export const watchErrorSummary = (errorCount: number): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Message,
    code: errorCount === 1 ? 6193 : 6194,
    messageText:
        errorCount === 1
            ? "Found 1 error. Watching for file changes."
            : `Found ${errorCount} errors. Watching for file changes.`,
});

const createCommandLineError = <TArgs extends any[]>(code: number, getMessage: (...args: TArgs) => string) =>
    createDiagnosticFactoryWithCode(code, (...args: TArgs) => ({ messageText: getMessage(...args) }));

export const unknownCompilerOption = createCommandLineError(
    5023,
    (name: string) => `Unknown compiler option '${name}'.`
);

export const compilerOptionRequiresAValueOfType = createCommandLineError(
    5024,
    (name: string, type: string) => `Compiler option '${name}' requires a value of type ${type}.`
);

export const optionProjectCannotBeMixedWithSourceFilesOnACommandLine = createCommandLineError(
    5042,
    () => "Option 'project' cannot be mixed with source files on a command line."
);

export const cannotFindATsconfigJsonAtTheSpecifiedDirectory = createCommandLineError(
    5057,
    (dir: string) => `Cannot find a tsconfig.json file at the specified directory: '${dir}'.`
);

export const theSpecifiedPathDoesNotExist = createCommandLineError(
    5058,
    (dir: string) => `The specified path does not exist: '${dir}'.`
);

export const compilerOptionExpectsAnArgument = createCommandLineError(
    6044,
    (name: string) => `Compiler option '${name}' expects an argument.`
);

export const argumentForOptionMustBe = createCommandLineError(
    6046,
    (name: string, values: string) => `Argument for '${name}' option must be: ${values}.`
);

export const optionCanOnlyBeSpecifiedInTsconfigJsonFile = createCommandLineError(
    6064,
    (name: string) => `Option '${name}' can only be specified in 'tsconfig.json' file.`
);

export const optionBuildMustBeFirstCommandLineArgument = createCommandLineError(
    6369,
    () => "Option '--build' must be the first command line argument."
);
