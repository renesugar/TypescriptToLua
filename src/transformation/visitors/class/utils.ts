import * as ts from "typescript";
import { TransformationContext } from "../../context";
import { AnnotationKind, getTypeAnnotations } from "../../utils/annotations";

export function isStaticNode(node: ts.Node): boolean {
    return (node.modifiers ?? []).some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendsClause(node: ts.ClassLikeDeclarationBase): ts.HeritageClause | undefined {
    return (node.heritageClauses ?? []).find(clause => clause.token === ts.SyntaxKind.ExtendsKeyword);
}

export function getExtendedNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    const extendsClause = getExtendsClause(node);
    if (!extendsClause) return;

    const superType = context.checker.getTypeAtLocation(extendsClause.types[0]);
    const annotations = getTypeAnnotations(superType);
    if (!annotations.has(AnnotationKind.PureAbstract)) {
        return extendsClause.types[0];
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedNode = getExtendedNode(context, node);
    return extendedNode && context.checker.getTypeAtLocation(extendedNode);
}
