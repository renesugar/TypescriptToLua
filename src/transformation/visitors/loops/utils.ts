import * as ts from "typescript";
import * as lua from "../../../LuaAST";
import { TransformationContext } from "../../context";
import { performHoisting, popScope, pushScope, ScopeType } from "../../utils/scope";
import { isAssignmentPattern } from "../../utils/typescript";
import { transformAssignment } from "../binary-expression/assignments";
import { transformAssignmentPattern } from "../binary-expression/destructuring-assignments";
import { transformBlockOrStatement } from "../block";
import { transformIdentifier } from "../identifier";
import { checkVariableDeclarationList, transformBindingPattern } from "../variable-declaration";

export function transformLoopBody(
    context: TransformationContext,
    loop: ts.WhileStatement | ts.DoStatement | ts.ForStatement | ts.ForOfStatement | ts.ForInOrOfStatement
): lua.Statement[] {
    pushScope(context, ScopeType.Loop);
    const body = performHoisting(context, transformBlockOrStatement(context, loop.statement));
    const scope = popScope(context);
    const scopeId = scope.id;

    if (!scope.loopContinued) {
        return body;
    }

    const baseResult: lua.Statement[] = [lua.createDoStatement(body)];
    const continueLabel = lua.createLabelStatement(`__continue${scopeId}`);
    baseResult.push(continueLabel);

    return baseResult;
}

export function getVariableDeclarationBinding(
    context: TransformationContext,
    node: ts.VariableDeclarationList
): ts.BindingName {
    checkVariableDeclarationList(context, node);

    if (node.declarations.length === 0) {
        return ts.createIdentifier("____");
    }

    return node.declarations[0].name;
}

export function transformForInitializer(
    context: TransformationContext,
    initializer: ts.ForInitializer,
    block: lua.Block
): lua.Identifier {
    const valueVariable = lua.createIdentifier("____value");

    if (ts.isVariableDeclarationList(initializer)) {
        // Declaration of new variable

        const binding = getVariableDeclarationBinding(context, initializer);
        if (ts.isArrayBindingPattern(binding) || ts.isObjectBindingPattern(binding)) {
            block.statements.unshift(...transformBindingPattern(context, binding, valueVariable));
        } else {
            // Single variable declared in for loop
            return transformIdentifier(context, binding);
        }
    } else {
        // Assignment to existing variable(s)

        block.statements.unshift(
            ...(isAssignmentPattern(initializer)
                ? transformAssignmentPattern(context, initializer, valueVariable)
                : transformAssignment(context, initializer, valueVariable))
        );
    }

    return valueVariable;
}
