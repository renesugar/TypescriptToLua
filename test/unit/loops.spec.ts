import * as tstl from "../../src";
import { forbiddenForIn, unsupportedForTarget } from "../../src/transformation/utils/diagnostics";
import * as util from "../util";

test("while", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        let i = 0;
        while (i < arrTest.length) {
            arrTest[i] = arrTest[i] + 1;
            i++;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("while with continue", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3, 4];
        let i = 0;
        while (i < arrTest.length) {
            if (i % 2 == 0) {
                i++;
                continue;
            }
            let j = 2;
            while (j > 0) {
                if (j == 2) {
                    j--
                    continue;
                }
                arrTest[i] = j;
                j--;
            }

            i++;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("dowhile with continue", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3, 4];
        let i = 0;
        do {
            if (i % 2 == 0) {
                i++;
                continue;
            }
            let j = 2;
            do {
                if (j == 2) {
                    j--
                    continue;
                }
                arrTest[i] = j;
                j--;
            } while (j > 0)

            i++;
        } while (i < arrTest.length)
        return arrTest;
    `.expectToMatchJsResult();
});

test("for", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        for (let i = 0; i < arrTest.length; ++i) {
            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("for with expression", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        let i: number;
        for (i = 0 * 1; i < arrTest.length; ++i) {
            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("for with continue", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3, 4];
        for (let i = 0; i < arrTest.length; i++) {
            if (i % 2 == 0) {
                continue;
            }

            for (let j = 0; j < 2; j++) {
                if (j == 1) {
                    continue;
                }
                arrTest[i] = j;
            }
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("forMirror", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        for (let i = 0; arrTest.length > i; i++) {
            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("forBreak", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        for (let i = 0; i < arrTest.length; ++i) {
            break;
            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("forNoDeclarations", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        let i = 0;
        for (; i < arrTest.length; ++i) {
            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("forNoCondition", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        let i = 0;
        for (;; ++i) {
            if (i >= arrTest.length) {
                break;
            }

            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("forNoPostExpression", () => {
    util.testFunction`
        let arrTest = [0, 1, 2, 3];
        let i = 0;
        for (;;) {
            if (i >= arrTest.length) {
                break;
            }

            arrTest[i] = arrTest[i] + 1;

            i++;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test.each([
    { inp: [0, 1, 2, 3], header: "let i = 0; i < arrTest.length; i++" },
    { inp: [0, 1, 2, 3], header: "let i = 0; i <= arrTest.length - 1; i++" },
    { inp: [0, 1, 2, 3], header: "let i = 0; arrTest.length > i; i++" },
    { inp: [0, 1, 2, 3], header: "let i = 0; arrTest.length - 1 >= i; i++" },
    { inp: [0, 1, 2, 3], header: "let i = 0; i < arrTest.length; i += 2" },
    { inp: [0, 1, 2, 3], header: "let i = arrTest.length - 1; i >= 0; i--" },
    { inp: [0, 1, 2, 3], header: "let i = arrTest.length - 1; i >= 0; i -= 2" },
    { inp: [0, 1, 2, 3], header: "let i = arrTest.length - 1; i > 0; i -= 2" },
])("forheader (%p)", ({ inp, header }) => {
    util.testFunction`
        let arrTest = ${util.formatCode(inp)};
        for (${header}) {
            arrTest[i] = arrTest[i] + 1;
        }
        return arrTest;
    `.expectToMatchJsResult();
});

test("for scope", () => {
    util.testFunction`
        let i = 42;
        for (let i = 0; i < 10; ++i) {}
        return i;
    `.expectToMatchJsResult();
});

test.each([
    {
        inp: { test1: 0, test2: 1, test3: 2 },
        expected: { test1: 1, test2: 2, test3: 3 },
    },
])("forin[Object] (%p)", ({ inp, expected }) => {
    const result = util.transpileAndExecute(
        `let objTest = ${JSON.stringify(inp)};
        for (let key in objTest) {
            objTest[key] = objTest[key] + 1;
        }
        return JSONStringify(objTest);`
    );

    expect(JSON.parse(result)).toEqual(expected);
});

test("forin[Array]", () => {
    util.testFunction`
        const array = [];
        for (const key in array) {}
    `.expectDiagnosticsToMatchSnapshot([forbiddenForIn.code]);
});

test.each([{ inp: { a: 0, b: 1, c: 2, d: 3, e: 4 }, expected: { a: 0, b: 0, c: 2, d: 0, e: 4 } }])(
    "forin with continue (%p)",
    ({ inp, expected }) => {
        const result = util.transpileAndExecute(
            `let obj = ${JSON.stringify(inp)};
            for (let i in obj) {
                if (obj[i] % 2 == 0) {
                    continue;
                }

                obj[i] = 0;
            }
            return JSONStringify(obj);`
        );

        expect(result).toBe(JSON.stringify(expected));
    }
);

test.each([{ inp: [0, 1, 2], expected: [1, 2, 3] }])("forof (%p)", ({ inp, expected }) => {
    const result = util.transpileAndExecute(
        `let objTest = ${JSON.stringify(inp)};
        let arrResultTest = [];
        for (let value of objTest) {
            arrResultTest.push(value + 1)
        }
        return JSONStringify(arrResultTest);`
    );

    expect(result).toBe(JSON.stringify(expected));
});

test("Tuple loop", () => {
    util.testFunction`
        const tuple: [number, number, number] = [3,5,1];
        let count = 0;
        for (const value of tuple) { count += value; }
        return count;
    `.expectToMatchJsResult();
});

test("forof existing variable", () => {
    util.testFunction`
        let objTest = [0, 1, 2];
        let arrResultTest = [];
        let value: number;
        for (value of objTest) {
            arrResultTest.push(value + 1)
        }
        return arrResultTest;
    `.expectToMatchJsResult();
});

test("forof destructing", () => {
    const input = [
        [1, 2],
        [2, 3],
        [3, 4],
    ];

    util.testFunction`
        let objTest = ${util.formatCode(input)};
        let arrResultTest = [];
        for (let [a,b] of objTest) {
            arrResultTest.push(a + b)
        }
        return arrResultTest;
    `.expectToMatchJsResult();
});

test("forof destructing with existing variables", () => {
    const input = [
        [1, 2],
        [2, 3],
        [3, 4],
    ];

    util.testFunction`
        let objTest = ${util.formatCode(input)};
        let arrResultTest = [];
        let a: number;
        let b: number;
        for ([a,b] of objTest) {
            arrResultTest.push(a + b)
        }
        return arrResultTest;
    `.expectToMatchJsResult();
});

test("forof with continue", () => {
    util.testFunction`
        let testArr = [0, 1, 2, 3, 4];
        let a = 0;
        for (let i of testArr) {
            if (i % 2 == 0) {
                a++;
                continue;
            }

            for (let j of [0, 1]) {
                if (j == 1) {
                    continue;
                }
                testArr[a] = j;
            }
            a++;
        }
        return testArr;
    `.expectToMatchJsResult();
});

test("forof with iterator", () => {
    util.testFunction`
        const arr = ["a", "b", "c"];
        function iter(): IterableIterator<string> {
            let i = 0;
            return {
                [Symbol.iterator]() { return this; },
                next() { return {value: arr[i], done: i++ >= arr.length} },
            }
        }
        let result = "";
        for (let e of iter()) {
            result += e;
        }
        return result;
    `.expectToMatchJsResult();
});

test("forof with iterator and existing variable", () => {
    util.testFunction`
        const arr = ["a", "b", "c"];
        function iter(): IterableIterator<string> {
            let i = 0;
            return {
                [Symbol.iterator]() { return this; },
                next() { return {value: arr[i], done: i++ >= arr.length} },
            }
        }
        let result = "";
        let e: string;
        for (e of iter()) {
            result += e;
        }
        return result;
    `.expectToMatchJsResult();
});

test("forof destructuring with iterator", () => {
    util.testFunction`
        const arr = ["a", "b", "c"];
        function iter(): IterableIterator<[string, string]> {
            let i = 0;
            return {
                [Symbol.iterator]() { return this; },
                next() { return {value: [i.toString(), arr[i]], done: i++ >= arr.length} },
            }
        }
        let result = "";
        for (let [a, b] of iter()) {
            result += a + b;
        }
        return result;
    `.expectToMatchJsResult();
});

test("forof destructuring with iterator and existing variables", () => {
    util.testFunction`
        const arr = ["a", "b", "c"];
        function iter(): IterableIterator<[string, string]> {
            let i = 0;
            return {
                [Symbol.iterator]() { return this; },
                next() { return {value: [i.toString(), arr[i]], done: i++ >= arr.length} },
            }
        }
        let result = "";
        let a: string;
        let b: string;
        for ([a, b] of iter()) {
            result += a + b;
        }
        return result;
    `.expectToMatchJsResult();
});

test("forof array which modifies length", () => {
    util.testFunction`
        const arr = ["a", "b", "c"];
        let result = "";
        for (const v of arr) {
            if (v === "a") {
                arr.push("d");
            }
            result += v;
        }
        return result;
    `.expectToMatchJsResult();
});

test("forof nested destructuring", () => {
    util.testFunction`
        let result = 0;
        for (const [[x]] of [[[1]]]) {
            result = x;
        }
        return result;
    `.expectToMatchJsResult();
});

test("forof with array typed as iterable", () => {
    util.testFunction`
        function foo(): Iterable<string> {
            return ["A", "B", "C"];
        }
        let result = "";
        for (const x of foo()) {
            result += x;
        }
        return result;
    `.expectToMatchJsResult();
});

test.each(["", "abc", "a\0c"])("forof string (%p)", string => {
    util.testFunctionTemplate`
        const results: string[] = [];
        for (const x of ${string}) {
            results.push(x);
        }
        return results;
    `.expectToMatchJsResult();
});

describe("for...of empty destructuring", () => {
    const declareTests = (destructuringPrefix: string) => {
        test("array", () => {
            const code = `
                const arr = [["a"], ["b"], ["c"]];
                let i = 0;
                for (${destructuringPrefix}[] of arr) {
                    ++i;
                }
                return i;
            `;
            expect(util.transpileAndExecute(code)).toBe(3);
        });

        test("iterable", () => {
            const code = `
                const iter: Iterable<string[]> = [["a"], ["b"], ["c"]];
                let i = 0;
                for (${destructuringPrefix}[] of iter) {
                    ++i;
                }
                return i;
            `;
            expect(util.transpileAndExecute(code)).toBe(3);
        });

        test("luaIterator", () => {
            const code = `
                const arr = [["a"], ["b"], ["c"]];
                /** @luaIterator */
                interface Iter extends Iterable<string[]> {}
                function luaIter(): Iter {
                    let it = 0;
                    return (() => arr[it++]) as any;
                }
                let i = 0;
                for (${destructuringPrefix}[] of luaIter()) {
                    ++i;
                }
                return i;
            `;
            expect(util.transpileAndExecute(code)).toBe(3);
        });

        test("luaIterator+tupleReturn", () => {
            const code = `
                const arr = [["a", "b"], ["c", "d"], ["e", "f"]];
                /**
                 * @luaIterator
                 * @tupleReturn
                 */
                interface Iter extends Iterable<[string, string]> {}
                function luaIter(): Iter {
                    let it = 0;
                    /** @tupleReturn */
                    function iter() {
                        const e = arr[it++];
                        if (e) {
                            return e;
                        }
                    }
                    return iter as any;
                }
                let i = 0;
                for (${destructuringPrefix}[] of luaIter()) {
                    ++i;
                }
                return i;
            `;
            expect(util.transpileAndExecute(code)).toBe(3);
        });
    };

    describe("declaration", () => declareTests("const "));
    describe("assignment", () => declareTests(""));
});

for (const testCase of [
    "while (false) { continue; }",
    "do { continue; } while (false)",
    "for (;;) { continue; }",
    "for (const a in {}) { continue; }",
    "for (const a of []) { continue; }",
]) {
    const expectContinueGotoLabel: util.TapCallback = builder =>
        expect(builder.getMainLuaCodeChunk()).toMatch("::__continue2::");

    util.testEachVersion(`loop continue (${testCase})`, () => util.testModule(testCase), {
        [tstl.LuaTarget.Universal]: builder => builder.expectDiagnosticsToMatchSnapshot([unsupportedForTarget.code]),
        [tstl.LuaTarget.Lua51]: builder => builder.expectDiagnosticsToMatchSnapshot([unsupportedForTarget.code]),
        [tstl.LuaTarget.Lua52]: expectContinueGotoLabel,
        [tstl.LuaTarget.Lua53]: expectContinueGotoLabel,
        [tstl.LuaTarget.LuaJIT]: expectContinueGotoLabel,
    });
}

test("do...while", () => {
    util.testFunction`
        let result = 0;
        do {
            ++result;
        } while (result < 2);
        return result;
    `.expectToMatchJsResult();
});

test("do...while scoping", () => {
    util.testFunction`
        let x = 0;
        let result = 0;
        do {
            let x = 1;
            ++result;
        } while (x === 0 && result < 2);
        return result;
    `.expectToMatchJsResult();
});

test("do...while double-negation", () => {
    const builder = util.testFunction`
        let result = 0;
        do {
            ++result;
        } while (!(result >= 2));
        return result;
    `.expectToMatchJsResult();

    expect(builder.getMainLuaCodeChunk()).not.toMatch("not");
});

test("for...in with pre-defined variable", () => {
    util.testFunction`
        const obj = { x: "y", foo: "bar" };

        let x = "";
        let result = [];
        for (x in obj) {
            result.push(x);
        }
        return result;
    `.expectToMatchJsResult();
});

test("for...in with pre-defined variable keeps last value", () => {
    util.testFunction`
        const obj = { x: "y", foo: "bar" };

        let x = "";
        for (x in obj) {
        }
        return x;
    `.expectToMatchJsResult();
});
