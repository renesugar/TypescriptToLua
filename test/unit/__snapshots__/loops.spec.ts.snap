// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forin[Array]: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local array = {}
    for key in pairs(array) do
    end
end
return ____exports"
`;

exports[`forin[Array]: diagnostics 1`] = `"main.ts(3,9): error TSTL: Iterating over arrays with 'for ... in' is not allowed."`;

exports[`loop continue (do { continue; } while (false)) [5.1]: code 1`] = `
"repeat
    do
        do
            goto __continue2
        end
        ::__continue2::
    end
until not false"
`;

exports[`loop continue (do { continue; } while (false)) [5.1]: diagnostics 1`] = `"main.ts(1,6): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (do { continue; } while (false)) [universal]: code 1`] = `
"repeat
    do
        do
            goto __continue2
        end
        ::__continue2::
    end
until not false"
`;

exports[`loop continue (do { continue; } while (false)) [universal]: diagnostics 1`] = `"main.ts(1,6): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (for (;;) { continue; }) [5.1]: code 1`] = `
"do
    while true do
        do
            goto __continue2
        end
        ::__continue2::
    end
end"
`;

exports[`loop continue (for (;;) { continue; }) [5.1]: diagnostics 1`] = `"main.ts(1,12): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (for (;;) { continue; }) [universal]: code 1`] = `
"do
    while true do
        do
            goto __continue2
        end
        ::__continue2::
    end
end"
`;

exports[`loop continue (for (;;) { continue; }) [universal]: diagnostics 1`] = `"main.ts(1,12): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (for (const a in {}) { continue; }) [5.1]: code 1`] = `
"for a in pairs({}) do
    do
        goto __continue2
    end
    ::__continue2::
end"
`;

exports[`loop continue (for (const a in {}) { continue; }) [5.1]: diagnostics 1`] = `"main.ts(1,23): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (for (const a in {}) { continue; }) [universal]: code 1`] = `
"for a in pairs({}) do
    do
        goto __continue2
    end
    ::__continue2::
end"
`;

exports[`loop continue (for (const a in {}) { continue; }) [universal]: diagnostics 1`] = `"main.ts(1,23): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (for (const a of []) { continue; }) [5.1]: code 1`] = `
"for ____, a in ipairs({}) do
    do
        goto __continue2
    end
    ::__continue2::
end"
`;

exports[`loop continue (for (const a of []) { continue; }) [5.1]: diagnostics 1`] = `"main.ts(1,23): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (for (const a of []) { continue; }) [universal]: code 1`] = `
"for ____, a in ipairs({}) do
    do
        goto __continue2
    end
    ::__continue2::
end"
`;

exports[`loop continue (for (const a of []) { continue; }) [universal]: diagnostics 1`] = `"main.ts(1,23): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (while (false) { continue; }) [5.1]: code 1`] = `
"while false do
    do
        goto __continue2
    end
    ::__continue2::
end"
`;

exports[`loop continue (while (false) { continue; }) [5.1]: diagnostics 1`] = `"main.ts(1,17): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;

exports[`loop continue (while (false) { continue; }) [universal]: code 1`] = `
"while false do
    do
        goto __continue2
    end
    ::__continue2::
end"
`;

exports[`loop continue (while (false) { continue; }) [universal]: diagnostics 1`] = `"main.ts(1,17): error TSTL: Continue statement is/are not supported for target Lua 5.1."`;
